Oauth Resource Owner Password / Client Credentials flows

nbarrera opened the pull request:
## Adds both oauth flows

### Resource Owner Password
```
     +----------+
     | Resource |
     |  Owner   |
     |          |
     +----------+
          v
          |    Resource Owner
         (A) Password Credentials
          |
          v
     +---------+                                  +---------------+
     |         |>--(B)---- Resource Owner ------->|               |
     |         |         Password Credentials     | Authorization |
     | Client  |                                  |     Server    |
     |         |<--(C)---- Access Token ---------<|               |
     |         |    (w/ Optional Refresh Token)   |               |
     +---------+                                  +---------------+
```


### Client Credentials
```
     +---------+                                  +---------------+
     |         |                                  |               |
     |         |>--(A)- Client Authentication --->| Authorization |
     | Client  |                                  |     Server    |
     |         |<--(B)---- Access Token ---------<|               |
     |         |                                  |               |
     +---------+                                  +---------------+
```

## How

Adds a new endpoint `/tokens` which is public.

Used `oauth2rize` lib which implements the process of exchanging some credentials for a new token, he will give you some callbacks to fill in the blanks but most of the job is tone by `oauth2rize`, then you just need to know how to use it. The work done by `oauth2rize` could be in a separate authentication server, that's why I just separated in a different file `auth-server.js`. Then we have more auth related stuff.. but that has to do with the actual api server security thing so it's in a different file `api-auth.js`.

So we ended up using `passport` to secure the api and `oauth2rize` to implement the auth server tasks.

## Considerations

Checkout the postman collection has changed, you will find the new `/token` request there in for the two `grant_types`.

## Pendings

- [x] Use private key to sign tokens / change algo to RS256?
- [x] Store password encrypted
- [ ] Fix code styling / run prettier?
- [ ] `passport` does not support promises, I didn't work on wrapping it's callbacks to use promises, so I ended up using more callback hell on my own code there 🤮 
- [x]  Add a way to create more clients through the API?
- [x] Check TODOs

nbarrera committed:
oauth addon first commit

using passport and oauth2orize

to authenticate and issue tokens

still needs more work on:

* create the token
* store the token
* define if the exchange /token endpoint will authenticate?

nbarrera committed:
adds jsonwebtoken and removes passprot-http

nbarrera committed:
adds token issuing

still needs work on..

correctly signing jwt

client credentials flow

nbarrera committed:
adding the /token endpoint on postman

also setting up Authorization header for whole collection

and no auth for /token

nbarrera committed:
securing all courses endpoints

nbarrera committed:
implemented client password and refactor

big move here, ended up without basic auth

/token is now public, added some ratelimt though

when hitting /token you could end using

the resource owner password exchange or the client credentials one

depending on what you 've sent on the body (grant_type)

each exchange will validate the input user / clientId

PENDING >>> refresh_token flow

nbarrera committed:
adding client credentials postman req

Javier-Rotelli commented:
tbh, this is working, we can fix the problems you identified, and perhaps run a couple of tests to ensure we aren't leaking anything. I'd say we go with passport.

nbarrera committed:
finished with the oauth part pendings

* impl signing using rsa key
* now passwords are stored hashed
* added new endpoint to create more clients
* addressed auth TODOs

nbarrera commented:
I 've implemented most pendings
I think it's worth to say that auth stuff is done

* impl signing using rsa key
* now passwords are stored hashed
* added new endpoint to create more clients
* addressed auth TODOs
